Add Repository card to search results when faceted on repository

drh-stanford opened the pull request:
This PR fixes #225. It adds a repository card when needed in the search results.

The PR adds a new finder for the Repository class:

```
repository = Arclight::Repository.find_by(name: 'My Repository')
repository = Arclight::Repository.find_by(slug: 'my-slug')
```

The PR does a few improvements on the repository test data too:
- Updates the `spec/fixtures/config/repositories.yml` with more realistic data
- Reorganizes the EAD fixture data to use repository slugs as directory names (ala the demo site)
- Indexes the seed data with the repositories.yml data (note that lib/generators has the real repo data and we might want to just use the template there and install the real repo data via the test_app_templates later)

![screen shot 2017-05-05 at 1 03 37 pm](https://cloud.githubusercontent.com/assets/1861171/25762366/59a13f54-3193-11e7-9e0d-76606b43ad09.png)

jkeck commented on lib/arclight/repository.rb:
```diff
@@ -59,11 +59,18 @@ def self.all
+    # @return [Repository]
+    def self.find_by(slug: nil, name: nil)
+      if slug
+        all.select { |repo| repo.slug == slug }.first
```
Can we use `find` here instead? (and then we can drop `first` at the end).  I'm surprised rubocop doesn't complain about this.

drh-stanford replied:
My understanding is that `find` is to grab the record by its "ID" value. `find_by` and `where` were the two options I found for a way to search by name or by slug. Or are you suggesting that we implement a `find` that looks for matches in both the slug and name data?

jkeck replied:
I mean using `all.find { |repo| repo.slug == slug }` instead of `all.select { |repo| repo.slug == slug }.first`

`select` iterates through the entire array that you call it on, and `find` will find the first and exit, which would make sense since we're simply calling `first` anyway.

jkeck replied:
`detect` could also be used (as they are the same method).

https://ruby-doc.org/core-2.2.3/Enumerable.html#method-i-find

drh-stanford replied:
ok gotcha

jkeck commented on app/helpers/arclight_helper.rb:
```diff
@@ -43,6 +43,14 @@ def fields_have_content?(document, field_accessor)
+  # @return [Repository]
+  def repository_faceted_on
+    repos = facets_from_request.find { |f| f.name == 'repository_sim' }.try(:items)
+    Arclight::Repository.find_by(name: repos.first.value) if repos.length == 1 && repos.first.value
```
If `try` fails here `repos` will be `nil`.  But we call `length` and `first` on repos on this line.  Could this be potentially problematic if we don't find a repository?

drh-stanford replied:
I can add that to the guard. I think it'd only happen if the repository facet was empty though, right? Maybe there's a smarter way to detect if we're faceted by repository?

jkeck commented:
Hmm, now that I look at the mockups, I'm not sure the "cards" from the repositories page is the exact same markup/layout that we'll want for the search results.  Perhaps we can get some clarification from @ggeisler.

ggeisler commented:
@jkeck @drh-stanford Yes, good eye Jessie. In the search results context it doesn't make as much sense to show the collection count and the learn more button so we left those off (p. 16 of the MVP set of mockups). It's a subtle thing but probably makes sense for the card to be slightly different here.

drh-stanford commented:
Ok, then it sounds like we can't reuse the partial for the repository card as-is. I'll update it...

jkeck commented on app/helpers/arclight_helper.rb:
```diff
@@ -43,6 +43,17 @@ def fields_have_content?(document, field_accessor)
+  #
+  # @return [Repository]
+  def repository_faceted_on
+    if try(:search_state) # rubocop: disable Style/GuardClause
```
If we want to follow the style guide w/o doing an inline disable here, we could add `return unless try(:search_state)` at the top of the method (which should be functionally equivalent).

drh-stanford commented:
ok, i believe i've addressed all the comments. thanks for the feedback. a note on the styling -- #222 still has a subtask to improve the styling, so I didn't do that in this PR.
